### PR TITLE
Make moe permute and final as custom op

### PR DESCRIPTION
**Pull request by limin2021**
# None:  Make the moe permute and final operations as independent custom ops

## Description

1. Make the moe permute and final operations as independent custom ops, so that is could be used in fused_moe python flow
2. Add CuteDslFusedMoE backend, which is composed of multiple python ops. And can be used to easily test newly developed/optimized group gemm ops in python flow.

## Test Coverage

UT:  

`pytest -s -o log_cli=true tests/unittest/_torch/modules/test_fused_moe.py -k "test_fused_moe_fp8_blockwise"`

`pytest -s -o log_cli=true tests/unittest/_torch/modules/test_fused_moe.py -k "test_fused_moe_fp8_blockwise_multi_gpu"`

Model UT:

`pytest -s -o log_cli=true "tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_cute_dsl_fp8_block_scales"`

`pytest -s -o log_cli=true "tests/integration/defs/accuracy/test_llm_api_pytorch.py::TestDeepSeekV3Lite::test_cute_dsl_fp8_block_scales_4gpus"`

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by djns99 — COMMENTED** [review]
This is an excellent change thank you, something I have been wanting to do for a long time!
Is there any chance we can make it so that `moe_kernels.cu` shares the same implementation? 
For example you are missing the optimisations from https://github.com/NVIDIA/TensorRT-LLM/pull/5215, and if the kernels are separate it will take a deliberate effort to get new features/optimisations

**Comment by limin2021** [comment]
> This is an excellent change thank you, something I have been wanting to do for a long time! Is there any chance we can make it so that `moe_kernels.cu` shares the same implementation? For example you are missing the optimisations from #5215, and if the kernels are separate it will take a deliberate effort to get new features/optimisations

Thanks for your good suggestion. As this MR is urgently needed by Barry. After discussing with yuxian, Barry and Tao Li, we would like to merge it first and then further refactor in the next MR (remove the duplicated logic in moe_kernels.cu and provide a better organization). Tracked by https://jirasw.nvidia.com/browse/TRTLLM-6150.

Do you think is it OK?

File: cpp/tensorrt_llm/kernels/moeUtilOp.cu [hunk 1]
```diff
@@ -0,0 +1,881 @@
...
+}
+
+template <int EXPERTS_PER_TOKEN, int LOG2_NUM_EXPERTS>
+bool fusedBuildExpertMapsSortFirstTokenBlockSize(int const* token_selected_experts,
+    int* unpermuted_token_selected_experts, int* permuted_source_token_ids, int64_t* expert_first_token_offset,
+    int64_t const num_tokens, int const num_experts_per_node, int const experts_per_token, int const start_expert,
+    int const end_expert, cudaStream_t stream)
+{
+    int const block_size = num_tokens;
+    if (num_tokens > 256)
```
Comment (litaotju) [inline cpp/tensorrt_llm/kernels/moeUtilOp.cu:206]: This is suprising. 

Is there a fallback path to support num tokens > 256?

**Review by litaotju — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+from .quantization import MoEWeightLoadingMode
+from .routing import BaseMoeRoutingMethod
+
+
+def swiglu_fused_moe(x):
+    x, gate = x.chunk(2, dim=-1)
+    return F.silu(gate) * x
+
+
+def cute_dsl_fp8_group_blockwise_gemm_ref(
```
Comment (litaotju) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py:20]: NIT. This should go to tests directory right?

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+            router_logits)
+        assert token_selected_experts.shape[
+            1] == self.routing_method.experts_per_token
+        assert token_selected_experts.shape == token_final_scales.shape
+        assert token_selected_experts.shape[0] == router_logits.shape[0]
+        assert token_final_scales.dtype == torch.float32
+        assert token_selected_experts.dtype == torch.int32
+
+        if self.apply_router_weight_on_input:
+            assert self.routing_method.top_k == 1, "Current workaround only supports top-1 routing"
```
Comment (litaotju) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py:160]: what is this? Is this the limitation of the cutlass backend or this new backend?

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+            assert x.dtype != torch.float8_e4m3fn, "Current workaround for apply_router_weight_on_input does not support fp8 input"
+            x = x * token_final_scales.to(x.dtype)
+            # TODO: remove this once we have correct fusedmoe kernel ready
+            token_final_scales = None
+
+        # quantize inputs
+        use_deepseek_fp8_block_scale = False
+        weight_dtype = self.w3_w1_weight.dtype
+        x_sf = None
+        if self.has_any_quant:
```
Comment (litaotju) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py:170]: We should only include the quant recipe supported by this DSL backend, and raise errors for others.

File: cpp/tensorrt_llm/thop/moeUtilOp.cpp [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+        "Tensor expert_first_token_offset_tensor, int hidden_size, int experts_per_token, int num_experts_per_node, "
+        "int tp_size, int tp_rank, int ep_size, int ep_rank, bool use_fp8_block_scaling)"
+        "-> (Tensor, Tensor, Tensor)");
+}
+
+TORCH_LIBRARY_IMPL(trtllm, CUDA, m)
+{
+    m.impl("moe_permute_op", &torch_ext::moe_permute_op);
+    m.impl("moe_finalize_scale_op", &torch_ext::run_moe_finalize_scale_op);
+    m.impl("moe_expand_op", &torch_ext::run_moe_expand_op);
```
Comment (litaotju) [inline cpp/tensorrt_llm/thop/moeUtilOp.cpp:448]: Is this op used?

**Review by djns99 — APPROVED** [review]

File: cpp/tensorrt_llm/kernels/moeUtilOp.cu (hunk 1 above)
Comment (limin2021) [reply cpp/tensorrt_llm/kernels/moeUtilOp.cu:206]: Yes. The logic is as follows: 
num tokens <= 256: call fusedBuildExpertMapsSortFirstToken
num tokens > 256, call buildExpertMaps->generateTokenPermutation

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py (hunk 2 above)
Comment (limin2021) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py:20]: It is needed for the overall flow now, especially for model accuracy test now. In the end, it will replaced with optimized cute_dsl kernels.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py (hunk 3 above)
Comment (limin2021) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py:160]: This logic is from fused_moe_cutlass.py. I just copy from it because our cute_dsl deepgemm flow should have the same logic as it.

File: cpp/tensorrt_llm/thop/moeUtilOp.cpp (hunk 5 above)
Comment (limin2021) [reply cpp/tensorrt_llm/thop/moeUtilOp.cpp:448]: moe_expand_op is not used now. It is a independent op whose logic has already been included in moe_permute_op. I use it for debug before. I want to remove it when all refactoring done especially when it really not needed.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py (hunk 4 above)
Comment (limin2021) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_cute_dsl.py:178]: Done.

File: cpp/tensorrt_llm/thop/moeUtilOp.cpp [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+    m.def(
+        "moe_permute_op(Tensor input, Tensor token_selected_experts, Tensor? token_final_scales, Tensor "
+        "fc1_expert_weights, Tensor fc2_expert_weights, Tensor[]? quant_scales, Tensor? input_sf, int "
+        "num_experts_on_rank, int tp_size, int tp_rank, int ep_size, int ep_rank, int cluster_size, int cluster_rank, "
+        "bool min_latency_mode, bool use_fp8_block_scaling)"
+        "-> (Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor, Tensor)");
+    m.def(
+        "moe_finalize_scale_op(Tensor gemm2_output, Tensor fc2_expert_biases, Tensor unpermuted_final_scales, Tensor "
+        "expanded_source_row_to_expanded_dest_row, Tensor expert_for_source_row, Tensor "
+        "expert_first_token_offset_tensor, int num_rows, int hidden_size, int experts_per_token, int "
```
[the remark this answers is not in the document]
Comment (limin2021) [reply cpp/tensorrt_llm/thop/moeUtilOp.cpp]: Thanks. Done.

**Review by Barry-Delaney — APPROVED** [review]

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py [hunk 7]
```diff
@@ -397,3 +397,82 @@ def _(
...
+        tp_rank: int,
+        ep_size: int,
+        ep_rank: int,
+    ):
+
+        return gemm2_output.new_empty((num_rows, hidden_size),
+                                      dtype=gemm2_output.dtype)
+
+
+def fp8_quantize_1x128(input: torch.Tensor):
```
[the remark this answers is not in the document]
Comment (liji-nv) [reply tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py:478]: You can just wrap it with a lambda for places calling it in test. Pickle is annoying.

**Comment by jeromeku** [comment]
@limin2021 

Confused by the naming of `CuteDslFusedMoE`.  AFAIK, it doesn't actually use [cuteDSL](https://docs.nvidia.com/cutlass/media/docs/pythonDSL/overview.html) to implement the GEMM kernel.  Is the intention to add a `cuteDSL` based implementation in the future, or is the implementation elsewhere in the repo?
